src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs

```diff
@@ -72,17 +73,20 @@ internal ZipArchiveEntry(ZipArchive archive, ZipCentralDirectoryFileHeader cd)
             _everOpenedForWrite = false;
             _outstandingWriteStream = null;
 
-            FullName = DecodeEntryName(cd.Filename);
+            FullName = DecodeBytesToString(cd.Filename);
 
             _lhUnknownExtraFields = null;
-            // the cd should have these as null if we aren't in Update mode
+            // the cd should have this as null if we aren't in Update mode
             _cdUnknownExtraFields = cd.ExtraFields;
-            _fileComment = cd.FileComment;
+
+            _fileComment = ZipHelper.GetEncodedTruncatedBytes(cd.FileComment, _archive.EntryNameAndCommentEncoding ?? Encoding.UTF8, ushort.MaxValue);
 
             _compressionLevel = null;
+
+            _hasUnicodeEntryNameOrComment = IsGeneralPurposeUnicodeFlagSet();
```

Comment:
Instead of maintaining `_hasUnicodeEntryNameOrComment`, can we just use `IsGeneralPurposeUnicodeFlagSet()` everywhere you are currently using the field?

I think the field is redundant and even in this PR in some cases you use one or the other. I think you can also get rid of it entirely.

Reply:
Based on this comment and the other ones, what I did was remove the method, and only use _generalPurposeBitFlag whenever we need to read an archive, or to write the bit value (in which case we need to check the boolean value to decide the bit value).

For all other tasks, we keep it simple with the new boolean.